SSA/Projects/InstCombine/LLVM/Semantics.lean

```diff
@@ -0,0 +1,283 @@
+ .some <| x ||| y
+
+
+/--
+The ‘or’ instruction returns the bitwise logical inclusive or of its two
+operands.
+-/
+def or? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+ .some <| x &&& y
+
+/--
+The ‘xor’ instruction returns the bitwise logical exclusive or of its two
+operands.  The xor is used to implement the “one’s complement” operation, which
+is the “~” operator in C.
+-/
+def xor? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+ .some <| x ^^^ y
+
+/--
+The value produced is the integer sum of the two operands.
+If the sum has unsigned overflow, the result returned is the mathematical result modulo 2n, where n is the bit width of the result.
+Because LLVM integers use a two’s complement representation, this instruction is appropriate for both signed and unsigned integers.
+-/
+def add? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+ .some <| x + y
+
+/--
+The value produced is the integer difference of the two operands.
+If the difference has unsigned overflow, the result returned is the mathematical result modulo 2n, where n is the bit width of the result.
+Because LLVM integers use a two’s complement representation, this instruction is appropriate for both signed and unsigned integers.
+-/
+def sub? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+ .some <| x - y
+
+/--
+The value produced is the integer product of the two operands.
+If the result of the multiplication has unsigned overflow, the result returned
+is the mathematical result modulo 2n, where n is the bit width of the result.
+Because LLVM integers use a two’s complement representation, and the result
+is the same width as the operands, this instruction returns the correct
+result for both signed and unsigned integers.
+
+If a full product (e.g. i32 * i32 -> i64) is needed, the operands should be
+sign-extended or zero-extended as appropriate to the width of the full product.
+-/
+def mul? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+ .some <| x * y
+
+/--
+The value produced is the unsigned integer quotient of the two operands.
+Note that unsigned integer division and signed integer division are distinct operations; for signed integer division, use ‘sdiv’.
+Division by zero is undefined behavior.
+-/
+def udiv? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+  match y.toNat with
+    | 0 => none
+    | _ => some $ BitVec.ofNat w (x.toNat / y.toNat)
+
+/--
+The value produced is the signed integer quotient of the two operands rounded towards zero.
+Note that signed integer division and unsigned integer division are distinct operations; for unsigned integer division, use ‘udiv’.
+Division by zero is undefined behavior.
+Overflow also leads to undefined behavior; this is a rare case, but can occur, for example, by doing a 32-bit division of -2147483648 by -1.
+-/
+def sdiv? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+  if y.toInt = 0
+  then none
+  else
+    let div := (x.toInt / y.toInt)
+    if div < Int.ofNat (2^w)
```

Comment:
Do we not need to check the other side of overflow as well?

```suggestion
    if div < Int.ofNat (2^w) && div >= - (Int.ofNat (2^w))
```

Reply:
same as above